poetry/repositories/pool.py

```diff
@@ -22,11 +22,11 @@ def __init__(
         if repositories is None:
             repositories = []
 
-        self._lookup: Dict[str, int] = {}
+        self._lookup: Dict[Optional[str], int] = {}
```

Comment:
Not sure why there is an `Optional` here -- None will all hash to the same type in the dict, and I don't think it should ever end up here. I suspect the error is found elsewhere.

Reply:
> One place where I have just updated the annotation to match the code but which smells funny to me is in pool.py. Is it really intended that the repository_name indexing self._lookup might be None? I tried asserting that it was not, but lots of testcases failed...

Reply:
I'm pretty sure this is unintentional behavior (even if it doesn't cause a bug). For now I'd just whitelist it with an annotation, and circle back to address this on a follow-up PR.

Reply:
I'm in two minds about this.  
- The advantage of changing the annotation as I have it is that we can then remove this file from the whitelist, and be confident that further typing errors will not happen.  
- The advantage of your suggestion is that it reminds us that there is a problem here.

Well, why not both?  Suggest that we leave the annotation in place for now - it does, after all, reflect what the code is actually doing - but also add a `# FIXME` saying "surely there's something wrong here".

(Again, I do not expect to follow-up with any fix for that - this looks like quite a pervasive change).

Reply:
I think the compromise here is good. Fair enough!